### PR TITLE
allow NILVALUE as pid in common prefix for logtype rfc5424

### PR DESCRIPTION
**Pull request by philhagen**
Before submitting your PR, please review the following checklist:

- [x] **CHOOSE CORRECT BRANCH**: if filing a bugfix/enhancement
      against certain release version, choose `0.9`, `0.10` or `0.11` branch,
      for dev-edition use `master` branch
- [ ] **CONSIDER adding a unit test** if your PR resolves an issue
- [x] **LIST ISSUES** this PR resolves
  - fixes part of #2711 
- [x] **MAKE SURE** this PR doesn't break existing tests
- [x] **KEEP PR small** so it could be easily reviewed.
- [x] **AVOID** making unnecessary stylistic changes in unrelated code

### CODE REVIEW

File: config/filter.d/common.conf [hunk 1]
```diff
@@ -83,7 +83,7 @@ datepattern = %(lt_short/datepattern)s
 [lt_rfc5424]
 # RFC 5424 log-format, see gh-2309:
 #__prefix_line = \s*<__hostname> <__daemon_re> \d+ \S+ \S+\s+
-__prefix_line = \s*<__hostname> <__daemon_re> \d+ \S+ (?:[^\[\]\s]+|(?:\[(?:[^\]"]*|"[^"]*")*\])+)\s+
+__prefix_line = \s*<__hostname> <__daemon_re> [\d+-] \S+ (?:[^\[\]\s]+|(?:\[(?:[^\]"]*|"[^"]*")*\])+)\s+
```
Comment (sebres) [inline config/filter.d/common.conf:86]: This is not correct RE, see https://github.com/fail2ban/fail2ban/issues/2711#issuecomment-623976515

**Comment by sebres** [comment]
I rebased your branch to current 0.10, but...

Although the standard tells about optional PID, we don't really want to follow the standards blindly.
As already said in #2711, the problem is that at the moment a common `__prefix_line` containing this is used as part of multi-line failure identifier (`<F-MLFID>`) in several filters.

For example such multi-line parsing filter is sshd, so a part of prefix `sshd 23602` is used there as an interim identifier (`<F-MLFID>`) to identify failures belonging to same session (IP). Furthermore such grouping allows to recognize success gained connection in order to forget other failures (acquiring for example by trying of several public keys), but also it allows to find some brute-force attempt targeting several user logins in same session (so in this case it would cause a more strict handling, for example don't allow reset of failed attempts counter), etc.

In any case such modification of common RE for rfc524 prefix would mistakenly prevent this identifying aims of the prefix-line, because it would stop to distinguish failures of different sessions anymore.

So I'd suggest = let us wait for the real example where such situation may occur in order to provide proper solution for this.
I'll block this PR unless it is occurred (and log excerpt is provided).

**Comment by philhagen** [comment]
Just re-tested this with fail2ban-0.11.1-9.el7.2.noarch (still CentOS defaults) and all lines are still missing the regex.

Redacted lines:
```
<86>1 2020-08-27T15:12:37.858815+00:00 simcoe sshd 19873 - -  reverse mapping checking getaddrinfo for host.example.com [1.2.3.4] failed - POSSIBLE BREAK-IN ATTEMPT!
<85>1 2020-08-27T15:12:37.924449+00:00 simcoe sshd 19873 - -  pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=1.2.3.4  user=root
<86>1 2020-08-27T15:12:37.924485+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:40.180599+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<86>1 2020-08-27T15:12:40.455772+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:42.791629+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<86>1 2020-08-27T15:12:43.059957+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:44.472637+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<86>1 2020-08-27T15:12:44.740647+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:46.624719+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<86>1 2020-08-27T15:12:46.892139+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:49.383734+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<86>1 2020-08-27T15:12:57.002466+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
<86>1 2020-08-27T15:12:58.866934+00:00 simcoe sshd 19873 - -  Failed password for root from 1.2.3.4 port 56589 ssh2
<83>1 2020-08-27T15:12:58.867058+00:00 simcoe sshd 19873 - -  error: maximum authentication attempts exceeded for root from 1.2.3.4 port 56589 ssh2 [preauth]
<86>1 2020-08-27T15:12:58.867084+00:00 simcoe sshd 19873 - -  Disconnecting: Too many authentication failures [preauth]
<85>1 2020-08-27T15:12:58.867964+00:00 simcoe sshd 19873 - -  PAM 5 more authentication failures; logname= uid=0 euid=0 tty=ssh ruser= rhost=1.2.3.4  user=root
<81>1 2020-08-27T15:12:58.867980+00:00 simcoe sshd 19873 - -  PAM service(sshd) ignoring max retries
```

Test command:
`fail2ban-regex -v -D /var/log/secure '/etc/fail2ban/filter.d/sshd.conf[logtype=rfc5424]'`

Results:
```
Running tests
=============

Use   failregex filter file : sshd, basedir: /etc/fail2ban
Use   filter options : {'logtype': 'rfc5424'}
Use         maxlines : 1
Use      datepattern : Default Detectors
Use         log file : /var/log/secure
Use         encoding : UTF-8


Results
=======

Failregex: 0 total
|-  #) [# of hits] regular expression
|   1) [0] ^[aA]uthentication (?:failure|error|failed) for <F-USER>.*</F-USER> from <HOST>( via \S+)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|   2) [0] ^User not known to the underlying authentication module for <F-USER>.*</F-USER> from <HOST>(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|   3) [0] ^Failed publickey for invalid user <F-USER>(?P<cond_user>\S+)|(?:(?! from ).)*?</F-USER> from <HOST>(?: (?:port \d+|on \S+)){0,2}(?: ssh\d*)?(?(cond_user): |(?:(?:(?! from ).)*)$)
|   4) [0] ^Failed \b(?!publickey)\S+ for (?P<cond_inv>invalid user )?<F-USER>(?P<cond_user>\S+)|(?(cond_inv)(?:(?! from ).)*?|[^:]+)</F-USER> from <HOST>(?: (?:port \d+|on \S+)){0,2}(?: ssh\d*)?(?(cond_user): |(?:(?:(?! from ).)*)$)
|   5) [0] ^<F-USER>ROOT</F-USER> LOGIN REFUSED FROM <HOST>
|   6) [0] ^[iI](?:llegal|nvalid) user <F-USER>.*?</F-USER> from <HOST>(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|   7) [0] ^User <F-USER>.+</F-USER> from <HOST> not allowed because not listed in AllowUsers(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|   8) [0] ^User <F-USER>.+</F-USER> from <HOST> not allowed because listed in DenyUsers(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|   9) [0] ^User <F-USER>.+</F-USER> from <HOST> not allowed because not in any group(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  10) [0] ^refused connect from \S+ \(<HOST>\)
|  11) [0] ^Received <F-MLFFORGET>disconnect</F-MLFFORGET> from <HOST>(?: (?:port \d+|on \S+)){0,2}:\s*3: .*: Auth fail(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  12) [0] ^User <F-USER>.+</F-USER> from <HOST> not allowed because a group is listed in DenyGroups(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  13) [0] ^User <F-USER>.+</F-USER> from <HOST> not allowed because none of user's groups are listed in AllowGroups(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  14) [0] ^<F-NOFAIL>pam_[a-z]+\(sshd:auth\):\s+authentication failure;</F-NOFAIL>(?:\s+(?:(?:logname|e?uid|tty)=\S*)){0,4}\s+ruser=<F-ALT_USER>\S*</F-ALT_USER>\s+rhost=<HOST>(?:\s+user=<F-USER>\S*</F-USER>)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  15) [0] ^(error: )?maximum authentication attempts exceeded for <F-USER>.*</F-USER> from <HOST>(?: (?:port \d+|on \S+)){0,2}(?: ssh\d*)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  16) [0] ^User <F-USER>.+</F-USER> not allowed because account is locked(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*
|  17) [0] ^<F-MLFFORGET>Disconnecting</F-MLFFORGET>(?: from)?(?: (?:invalid|authenticating)) user <F-USER>\S+</F-USER> <HOST>(?: (?:port \d+|on \S+)){0,2}:\s*Change of username or service not allowed:\s*.*\[preauth\]\s*$
|  18) [0] ^<F-MLFFORGET>Disconnecting</F-MLFFORGET>: Too many authentication failures(?: for <F-USER>.+?</F-USER>)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  19) [0] ^<F-NOFAIL>Received <F-MLFFORGET>disconnect</F-MLFFORGET></F-NOFAIL> from <HOST>(?: (?:port \d+|on \S+)){0,2}:\s*11:
|  20) [0] ^<F-NOFAIL><F-MLFFORGET>(Connection closed|Disconnected)</F-MLFFORGET></F-NOFAIL> (?:by|from)(?: (?:invalid|authenticating) user <F-USER>\S+|.+?</F-USER>)? <HOST>(?:(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*|\s*)$
|  21) [0] ^<F-MLFFORGET><F-MLFGAINED>Accepted \w+</F-MLFGAINED></F-MLFFORGET> for <F-USER>\S+</F-USER> from <HOST>(?:\s|$)
|  22) [0] ^<F-NOFAIL>Connection from</F-NOFAIL> <HOST>
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [27544] ^<\d+>\d+\s+ExYear(?P<_sep>[-/.])Month(?P=_sep)Day(?:T|  ?)24hour:Minute:Second(?:[.,]Microseconds)?(?:\s*Zone offset)?
|  [0] ^<\d+>\d+\s+(?:DAY )?MON Day %k:Minute:Second(?:\.Microseconds)?(?: ExYear)?
|  [0] ^<\d+>\d+\s+(?:DAY )?MON Day ExYear %k:Minute:Second(?:\.Microseconds)?
|  [0] ^<\d+>\d+\s+Day(?P<_sep>[-/])Month(?P=_sep)(?:ExYear|ExYear2) %k:Minute:Second
|  [0] ^<\d+>\d+\s+Day(?P<_sep>[-/])MON(?P=_sep)ExYear[ :]?24hour:Minute:Second(?:\.Microseconds)?(?: Zone offset)?
|  [0] ^<\d+>\d+\s+Month/Day/ExYear:24hour:Minute:Second
|  [0] ^<\d+>\d+\s+Month-Day-ExYear %k:Minute:Second(?:\.Microseconds)?
|  [0] ^<\d+>\d+\s+Epoch
|  [0] ^<\d+>\d+\s+24hour:Minute:Second
|  [0] ^<\d+>\d+\s+^<Month/Day/ExYear2@24hour:Minute:Second>
|  [0] ^<\d+>\d+\s+ExYear2ExMonthExDay  ?24hour:Minute:Second
|  [0] ^<\d+>\d+\s+MON Day, ExYear 12hour:Minute:Second AMPM
|  [0] ^<\d+>\d+\s+^MON-Day-ExYear2 %k:Minute:Second
|  [0] ^<\d+>\d+\s+ExYearExMonthExDay(?:T|  ?)Ex24hourExMinuteExSecond(?:[.,]Microseconds)?(?:\s*Zone offset)?
|  [0] ^<\d+>\d+\s+(?:Zone name )?(?:DAY )?MON Day %k:Minute:Second(?:\.Microseconds)?(?: ExYear)?
|  [0] ^<\d+>\d+\s+(?:Zone offset )?(?:DAY )?MON Day %k:Minute:Second(?:\.Microseconds)?(?: ExYear)?
|  [0] ^<\d+>\d+\s+TAI64N
`-

Lines: 27544 lines, 0 ignored, 0 matched, 27544 missed
[processed in 2.72 sec]

Missed line(s): too many to print.  Use --print-all-missed to print all 27544 lines
```

**Comment by sebres** [comment]
> and all lines are still missing the regex

This has nothing to do with NULL as PID (rather it is continuation of #2711).
And as you can see since https://github.com/fail2ban/fail2ban/blob/62b1712d223269b6201366e55c796a15a2f2211d/fail2ban/tests/files/logs/sshd#L350-L351
the format belongs to test-suite and tests are passing. And this (62b1712d223269b6201366e55c796a15a2f2211d) is already released with 0.11.1.
It is still a mystery to me why it does not work with 0.11.1-9.el7.2.noarch. If I test this with released stock 0.11.1 it works well.

You can also do it by yourself (for details see in [our wiki](https://github.com/fail2ban/fail2ban/wiki/How-to-test-newer-fail2ban-version-resp.-use-fail2ban-standalone-instance) how you can test other version without to install it):
<details><summary>PoC (test fail2ban-regex with stock 0.11.1)</summary>

```
$ cd /tmp
$ git clone -b 0.11.1 --single-branch https://github.com/fail2ban/fail2ban.git /tmp/f2b11.1
$ cd /tmp/f2b11.1
$ PYTHONPATH=. python bin/fail2ban-regex -c "$(pwd)/config" /tmp/gh-2711.txt sshd[logtype=rfc5424]

Running tests
=============

Use   failregex filter file : sshd, basedir: /tmp/f2b11.1/config
Use   filter options : {'logtype': 'rfc5424'}
Use         maxlines : 1
Use      datepattern : Default Detectors
Use         log file : /tmp/gh-2711.txt
Use         encoding : UTF-8


Results
=======

Failregex: 9 total
|-  #) [# of hits] regular expression
|   4) [6] ^Failed \b(?!publickey)\S+ for (?P<cond_inv>invalid user )?<F-USER>(?P<cond_user>\S+)|(?(cond_inv)(?:(?! from ).)*?|[^:]+)</F-USER> from <HOST>(?: (?:port \d+|on \S+)){0,2}(?: ssh\d*)?(?(cond_user): |(?:(?:(?! from ).)*)$)
|  14) [1] ^<F-NOFAIL>pam_[a-z]+\(sshd:auth\):\s+authentication failure;</F-NOFAIL>(?:\s+(?:(?:logname|e?uid|tty)=\S*)){0,4}\s+ruser=<F-ALT_USER>\S*</F-ALT_USER>\s+rhost=<HOST>(?:\s+user=<F-USER>\S*</F-USER>)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  15) [1] ^(error: )?maximum authentication attempts exceeded for <F-USER>.*</F-USER> from <HOST>(?: (?:port \d+|on \S+)){0,2}(?: ssh\d*)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
|  18) [1] ^<F-MLFFORGET>Disconnecting</F-MLFFORGET>: Too many authentication failures(?: for <F-USER>.+?</F-USER>)?(?: (?:port \d+|on \S+|\[preauth\])){0,3}\s*$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [18] ^<\d+>\d+\s+ExYear(?P<_sep>[-/.])Month(?P=_sep)Day(?:T|  ?)24hour:Minute:Second(?:[.,]Microseconds)?(?:\s*Zone offset)?
`-

Lines: 18 lines, 1 ignored, 8 matched, 9 missed
[processed in 0.00 sec]

|- Ignored line(s):
|  <85>1 2020-08-27T15:12:37.924449+00:00 simcoe sshd 19873 - -  pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=1.2.3.4  user=root
`-
|- Missed line(s):
|  <86>1 2020-08-27T15:12:37.858815+00:00 simcoe sshd 19873 - -  reverse mapping checking getaddrinfo for host.example.com [1.2.3.4] failed - POSSIBLE BREAK-IN ATTEMPT!
|  <86>1 2020-08-27T15:12:37.924485+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <86>1 2020-08-27T15:12:40.455772+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <86>1 2020-08-27T15:12:43.059957+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <86>1 2020-08-27T15:12:44.740647+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <86>1 2020-08-27T15:12:46.892139+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <86>1 2020-08-27T15:12:57.002466+00:00 simcoe sshd 19873 - -  pam_succeed_if(sshd:auth): requirement "uid >= 500" not met by user "root"
|  <85>1 2020-08-27T15:12:58.867964+00:00 simcoe sshd 19873 - -  PAM 5 more authentication failures; logname= uid=0 euid=0 tty=ssh ruser= rhost=1.2.3.4  user=root
|  <81>1 2020-08-27T15:12:58.867980+00:00 simcoe sshd 19873 - -  PAM service(sshd) ignoring max retries
`-
```
</details>

Have you done some customizations?
I'm also unsure whether some distribution-related changes have been made by @fail2ban/maintainers in config for noarch.
To find what is different or wrong, please define a single jail for sshd like below and provide an excerpt output of your `fail2ban-client -d` or `fail2ban-client --dp` for this jail:
```ini
[sshd]
filter = sshd[logtype=rfc5424]
enabled = true:
```
And you can compare it with this output of stock version:
<details><summary>PoC (dump of sshd, logtype=rfc5424 with stock 0.11.1)</summary>

```
/tmp/f2b11.1$ printf '[sshd]\nfilter = sshd[logtype=rfc5424]\nenabled = true\n' > /tmp/f2b11.1/config/jail.local
/tmp/f2b11.1$ PYTHONPATH=. python bin/fail2ban-client -c "$(pwd)/config" --dp
...
['add', 'sshd', 'auto']
['set', 'sshd', 'usedns', 'warn']
['set', 'sshd', 'prefregex', '^<F-MLFID>\\s*\\S+ [\\[\\(]?sshd(?:\\(\\S+\\))?[\\]\\)]?:? \\d+ \\S+ (?:[^\\[\\]\\s]+|(?:\\[(?:[^\\]"]*|"[^"]*")*\\])+)\\s+</F-MLFID>(?:(?:error|fatal): (?:PAM: )?)?<F-CONTENT>.+</F-CONTENT>$']
['set', 'sshd', 'maxlines', 1]
[ 'multi-set',
  'sshd',
  'addfailregex',
  [
  ...
    '^Failed \\b(?!publickey)\\S+ for (?P<cond_inv>invalid user )?<F-USER>(?P<cond_user>\\S+)|(?(cond_inv)(?:(?! from ).)*?|[^:]+)</F-USER> from <HOST>(?: (?:port \\d+|on \\S+)){0,2}(?: ssh\\d*)?(?(cond_user): |(?:(?:(?! from ).)*)$)',
  ...
    '^<F-NOFAIL>pam_[a-z]+\\(sshd:auth\\):\\s+authentication failure;</F-NOFAIL>(?:\\s+(?:(?:logname|e?uid|tty)=\\S*)){0,4}\\s+ruser=<F-ALT_USER>\\S*</F-ALT_USER>\\s+rhost=<HOST>(?:\\s+user=<F-USER>\\S*</F-USER>)?(?: (?:port \\d+|on \\S+|\\[preauth\\])){0,3}\\s*$',
  ...
    '^<F-MLFFORGET>Disconnecting</F-MLFFORGET>: Too many authentication failures(?: for <F-USER>.+?</F-USER>)?(?: (?:port \\d+|on \\S+|\\[preauth\\])){0,3}\\s*$',
    '^<F-NOFAIL>Received <F-MLFFORGET>disconnect</F-MLFFORGET></F-NOFAIL> from <HOST>(?: (?:port \\d+|on \\S+)){0,2}:\\s*11:',
    '^<F-NOFAIL><F-MLFFORGET>(Connection closed|Disconnected)</F-MLFFORGET></F-NOFAIL> (?:by|from)(?: (?:invalid|authenticating) user <F-USER>\\S+|.+?</F-USER>)? <HOST>(?:(?: (?:port \\d+|on \\S+|\\[preauth\\])){0,3}\\s*|\\s*)$',
    '^<F-MLFFORGET><F-MLFGAINED>Accepted \\w+</F-MLFGAINED></F-MLFFORGET> for <F-USER>\\S+</F-USER> from <HOST>(?:\\s|$)',
    '^<F-NOFAIL>Connection from</F-NOFAIL> <HOST>'
  ...
  ]]
['set', 'sshd', 'datepattern', '^<\\d+>\\d+\\s+{DATE}']
['set', 'sshd', 'logencoding', 'auto']
...
```
</details>

**Comment by philhagen** [comment]
I finally had an opportunity to revisit this and it seems things are working now.  I cleared some `.local` sections that I had not previously realized were in play and the banning is working as intended. Thank you for your assistance in tracking down the issue and I apologize for taking up so much of your time!

**Pull request closed without merging**